File: dht.js
var nodes = [];

var buckets = [];

//For now, I'm hard-coding the bucket list size.
//In the future, it should be customizable, but always a power of two
//to make math easier
function initDHT(our_id)
{
    var temp_id = Uint8Array(our_id);
    //First ideal bucket is us, other ideals are based on our ID
   buckets.push({ideal:our_id, nodes:[]});
    
    //Space our ideal ID's apart by 1/32 of the ID space
    var id_space_fraction = idBitShiftRight(getMaxId(), 5);
    for(var i = 0; i < 31; ++i)
    {
        temp_id = id_add(temp_id, id_space_fraction);
        buckets.push({ideal:temp_id, nodes:[]});
    }
    
    buckets.sort(function(a, b) {return idCompare(a.ideal, b.ideal);});
}

//Here's what we need to do: create a "distance metric" to each ideal ID
//The nodes then go to the bucket of the ID with the lowest distance metric
//I'm thinking it would be abs(ideal - nodeid)
//Or better, ((ideal > nodeid) ? ideal - nodeid : nodeid - ideal)
//^ better since we're dealing with unsigned integers

function getLeastDistanceIdealIndex(node_array)
{
    var lowest = getMaxId();
    var lowest_index = 0;
    
    for(var i = 0; i < buckets.length; ++i)
    {
        if(idCompare(lowest, idDistance(buckets[i].ideal, node_array)) > 0)
        {
            lowest_index = i;
            lowest = buckets[i].ideal;
        }
    }
    
    return lowest_index;
}

function addNode(node_id, node_channel)
{
    var node_array = parseIdFromHexString(node_id);
    
    buckets[getLeastDistanceIdealIndex(node_array)].nodes.push({id:node_array, channel:node_channel});
}

function getClosestNodeChannel(node_id)
{
    var node_array = parseIdFromHexString(node_id);
    var lowest = getMaxId();
    var lowestchannel;
    
    for(var i = 0; i < buckets.length; ++i)
    {
        for(var j = 0; j < buckets[i].nodes.length; ++j)
        {
            if(idCompare(lowest, idDistance(buckets[i].nodes[j].id, node_array)) > 0)
            {
                lowest = buckets[i].nodes[j].id;
                lowestchannel = buckets[i].nodes[j].channel;
            }
        }
    }
    
    return lowestchannel;
}

File: server/dht.c
#include <stdlib.h>

int ideal_bucket_compare(const void *a, const void *b)
{
    return id_compare((*(IdealBucket*)a).ideal, (*(IdealBucket*)b).ideal);
}

void init_DHT(BucketList &bl, ID our_id)
{
    ID temp = our_id;
    ID id_space_fraction = id_bit_shift_right(get_max_ID(), IDEAL_BUCKET_POWER_OF_TWO);
    int i = 0;
    for(i; i < NUM_IDEAL_BUCKETS; ++i)
    {
        bl.ideal_buckets[i].ideal = temp;
        bl.ideal_buckets[i].numbuckets = 0;
        bl.ideal_buckets[i].capacity = 0;
        temp = id_add(temp, id_space_fraction);
    }
    
    qsort(bl.ideal_buckets, NUM_IDEAL_BUCKETS, sizeof(IdealBucket), ideal_bucket_compare);
}

int get_least_distance_ideal_index(BucketList &bl, ID &node_id)
{
    ID lowest = getMaxId();
    int lowest_index = 0;
    int i = 0;
    
    for(i; i < NUM_IDEAL_BUCKETS; ++i)
    {
        if(id_compare(lowest, id_distance(bl.ideal_buckets[i].ideal, node_id)) > 0)
        {
            lowest_index = i;
            lowest = bl.ideal_buckets[i].ideal;
        }
    }
    
    return lowest_index;
}

int bucket_compare(const void *a, const void *b)
{
    return id_compare((*(Bucket*)a).id, (*(Bucket*)b).id);
}

void add_node(BucketList &bl, char* node_id_string, libwebsock_client_state *node_channel)
{
    ID node_id = parse_id_from_hex_string(node_id_string);
    
    int ideal_index = get_least_distance_ideal_index(bl, node_id);
    
    Bucket *temp;
    
    //Really don't want to type that a bunch of times, looks ugly
    Bucket *lessbullshit = &(bl.ideal_buckets[ideal_index]);
    unsigned int newcapacity;
    
    //Make sure we have enough space for a new bucket
    if(lessbullshit->capacity < (lessbullshit->numbuckets + 1))
    {
        newcapacity = (lessbullshit->capacity != 0) ? 2 * lessbullshit->capacity : 1;
        temp = malloc(sizeof(Bucket) * newcapacity);
        memcpy(temp, lessbullshit->buckets, lessbullshit->numbuckets * sizeof(Bucket));
        lessbullshit->capacity = newcapacity;
    }
    
    lessbullshit->buckets[lessbullshit->numbuckets++].id = node_id;
    lessbullshit->buckets[lessbullshit->numbuckets].client = node_channel;
    
    qsort(lessbullshit->buckets, lessbullshit->numbuckets, sizeof(Bucket), bucket_compare);
}

libwebsock_client_state* get_closest_node_channel(BucketList &bl, char *node_id_string)
{
    ID node_id = parse_id_from_hex_string(node_id_string);
    ID lowest = get_max_ID();
    libwebsock_client_state* lowestchannel;
    
    int i = 0;
    int j;
    //TODO: This should really be something like a binary search rather than a linear search, since the buckets are sorted
    for(i; i < NUM_IDEAL_BUCKETS; ++i)
    {
        for(j = 0; j < bl.ideal_buckets[i].numbuckets; ++j)
        {
            if(id_compare(lowest, id_distance(bl.ideal_buckets[i].buckets[j].id, node_id)) > 0)
            {
                lowest = bl.ideal_buckets[i].buckets[j].id;
                lowestchannel = bl.ideal_buckets[i].buckets[j].client;
            }
        }
    }
    
    return lowestchannel;
}
